### PR TITLE
Redesign public API for templates

### PR DESCRIPTION
**Pull request by rynowak**
-Renamed RouteTemplate -> RoutePattern
-Made immutable
-Added Builder
-Lots of fixes to parser to support new design

There are a few small issues logged for follow-up but this is mostly in
the place I want it design-wise.

### CODE REVIEW

**Comment by rynowak** [comment]
@jbagga please have a look

File: src/Microsoft.AspNetCore.Dispatcher/Patterns/InlineRouteParameterParser.cs [hunk 1]
```diff
@@ -3,12 +3,13 @@
 
 using System;
 using System.Collections.Generic;
+using System.Linq;
 
-namespace Microsoft.AspNetCore.Dispatcher
+namespace Microsoft.AspNetCore.Dispatcher.Patterns
 {
     public static class InlineRouteParameterParser
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Dispatcher/Patterns/InlineRouteParameterParser.cs]: This class is totally broken. Everything here should be a detail of the parser. I logged #475 to follow up on this

File: src/Microsoft.AspNetCore.Dispatcher/Patterns/ConstraintReference.cs [hunk 2]
```diff
@@ -0,0 +1,45 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Diagnostics;
+
+namespace Microsoft.AspNetCore.Dispatcher.Patterns
+{
+    /// <summary>
+    /// The parsed representation of an constraint in a <see cref="RoutePattern"/> parameter.
```
Comment (jbagga) [inline src/Microsoft.AspNetCore.Dispatcher/Patterns/ConstraintReference.cs]: nit: "a constraint"

File: src/Microsoft.AspNetCore.Dispatcher/Patterns/InlineRouteParameterParser.cs [hunk 3]
```diff
@@ -81,19 +75,15 @@ public static TemplatePart ParseRouteParameter(string routeParameter)
...
+            return new RoutePatternParameter(text, parameterName, defaultValue, parameterKind, parseResults.Constraints.ToArray());
         }
 
         private static ConstraintParseResults ParseConstraints(
             string routeParameter,
             int currentIndex,
             int endIndex)
         {
-            var inlineConstraints = new List<InlineConstraint>();
+            var inlineConstraints = new List<ConstraintReference>();
```
Comment (jbagga) [inline src/Microsoft.AspNetCore.Dispatcher/Patterns/InlineRouteParameterParser.cs]: Should this variable be renamed to something else? Or is `inlineConstraints` still relevant?

File: src/Microsoft.AspNetCore.Dispatcher/Patterns/InlineRouteParameterParser.cs [hunk 4]
```diff
@@ -126,7 +116,7 @@ public static TemplatePart ParseRouteParameter(string routeParameter)
                             case null:
                                 state = ParseState.End;
                                 var constraintText = routeParameter.Substring(startIndex, currentIndex - startIndex);
-                                inlineConstraints.Add(new InlineConstraint(constraintText));
+                                inlineConstraints.Add(ConstraintReference.Create(constraintText));
```
Comment (jbagga) [inline src/Microsoft.AspNetCore.Dispatcher/Patterns/InlineRouteParameterParser.cs]: What's the benefit of this pattern of .Create vs the ctor/new pattern?

File: src/Microsoft.AspNetCore.Dispatcher/Patterns/RoutePatternPart.cs [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+        }
+
+        public static RoutePatternLiteral CreateLiteral(string content)
+        {
+            if (string.IsNullOrEmpty(content))
+            {
+                throw new ArgumentException(Resources.Argument_NullOrEmpty, nameof(content));
+            }
+
+            if (content.IndexOf('?') >= 0)
```
Comment (jbagga) [inline src/Microsoft.AspNetCore.Dispatcher/Patterns/RoutePatternPart.cs]: Why is this looking for '?' specifically?

File: src/Microsoft.AspNetCore.Dispatcher/Patterns/RoutePatternPart.cs [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+        }
+
+        public static RoutePatternParameter CreateParameter(
+            string rawText,
+            string name,
+            object defaultValue,
+            RoutePatternParameterKind parameterKind,
+            params ConstraintReference[] constraints)
+        {
+
```
Comment (jbagga) [inline src/Microsoft.AspNetCore.Dispatcher/Patterns/RoutePatternPart.cs]: nit: empty line

File: src/Microsoft.AspNetCore.Dispatcher/Patterns/InlineRouteParameterParser.cs (hunk 3 above)
Comment (rynowak) [reply src/Microsoft.AspNetCore.Dispatcher/Patterns/InlineRouteParameterParser.cs]: yeah that's a good point, there are probably plenty of variable names to clean up

File: src/Microsoft.AspNetCore.Dispatcher/Patterns/InlineRouteParameterParser.cs (hunk 4 above)
Comment (rynowak) [reply src/Microsoft.AspNetCore.Dispatcher/Patterns/InlineRouteParameterParser.cs]: There are two reasons:
- Gives you the flexibility to return a subclass in the future if the design changes
- Can come up with more meaningful names (Look at update) when there are multiple ways to create something
- Consistency with one of the other parts of the system that does similar things, but for a better reason

File: src/Microsoft.AspNetCore.Dispatcher/Patterns/RoutePatternPart.cs (hunk 5 above)
Comment (rynowak) [reply src/Microsoft.AspNetCore.Dispatcher/Patterns/RoutePatternPart.cs]: `?`  is the delimiter for a query string - you can't put that in a path segment

File: src/Microsoft.AspNetCore.Dispatcher/Patterns/RoutePatternPart.cs (hunk 6 above)
Comment (rynowak) [reply src/Microsoft.AspNetCore.Dispatcher/Patterns/RoutePatternPart.cs]: 👍

**Pull request merged**